tools/outline_proxy_controller/outline_controller_server.cpp

```diff
@@ -94,51 +94,54 @@ std::tuple<int, std::string, std::string> session::runClientCommand(std::string
-    return std::make_tuple(GENERIC_FAILURE, "Invalid JSON", "");
+    return {static_cast<int>(ErrorCode::kUnexpected), "Invalid JSON", {}};
   }
   action = boost::lexical_cast<std::string>(pt.to_iterator(_action_iter)->second.data());
   // std::cout << action << std::endl;
 
-  if (action == CONFIGURE_ROUTING) {
-    boost::property_tree::ptree::assoc_iterator _parameters_iter = pt.find("parameters");
-    if (_parameters_iter == pt.not_found()) {
-      std::cerr << "Invalid input JSON - parameters doesn't exist" << std::endl;
-      return std::make_tuple(GENERIC_FAILURE, "Invalid JSON", action);
+  try {
+    if (action == CONFIGURE_ROUTING) {
+      boost::property_tree::ptree::assoc_iterator _parameters_iter = pt.find("parameters");
+      if (_parameters_iter == pt.not_found()) {
+        std::cerr << "Invalid input JSON - parameters doesn't exist" << std::endl;
+        return {static_cast<int>(ErrorCode::kUnexpected), "Invalid JSON", action};
+      }
+      boost::property_tree::ptree parameters = pt.to_iterator(_parameters_iter)->second;
+      boost::property_tree::ptree::assoc_iterator _proxyIp_iter = parameters.find("proxyIp");
+      if (_proxyIp_iter == parameters.not_found()) {
+        std::cerr << "Invalid input JSON - parameters doesn't exist" << std::endl;
+        return {static_cast<int>(ErrorCode::kUnexpected), "Invalid JSON", action};
+      }
+      outline_server_ip =
+          boost::lexical_cast<std::string>(pt.to_iterator(_proxyIp_iter)->second.data());
+
+      outlineProxyController_->routeThroughOutline(outline_server_ip);
+      std::cout << "Configure Routing to " << outline_server_ip << " is done." << std::endl;
+      return {static_cast<int>(ErrorCode::kOk), {}, action};
+    } else if (action == RESET_ROUTING) {
+      outlineProxyController_->routeDirectly();
+      std::cout << "Reset Routing done" << std::endl;
+      return {static_cast<int>(ErrorCode::kOk), {}, action};
+    } else if (action == GET_DEVICE_NAME) {
+      std::cout << "Get device name done" << std::endl;
+      return {static_cast<int>(ErrorCode::kOk), outlineProxyController_->getTunDeviceName(), action};
+    } else {
+      std::cerr << "Invalid action specified in JSON (" << action << ")" << std::endl;
+      return {static_cast<int>(ErrorCode::kUnexpected), "Undefined Action", {}};
     }
-    boost::property_tree::ptree parameters = pt.to_iterator(_parameters_iter)->second;
-    boost::property_tree::ptree::assoc_iterator _proxyIp_iter = parameters.find("proxyIp");
-    if (_proxyIp_iter == parameters.not_found()) {
-      std::cerr << "Invalid input JSON - parameters doesn't exist" << std::endl;
-      return std::make_tuple(GENERIC_FAILURE, "Invalid JSON", action);
+  } catch (const std::system_error& err) {
+    std::cerr << "[" << err.code() << "] " << err.what() << std::endl;
+    if (err.code().category() == OutlineErrorCategory()) {
+      return {err.code().value(), {}, action};
```

Comment:
Perhaps we should add the err.what to the return value too, both here and below.

Reply:
I have thought about it. The only concern here is we won't escape that in json string, so if we have quotation marks in `e.what()`, the response is not a valid json.

Reply:
That's a serious issue that could expose us to vulnerability. We should be using a proper JSON library and not assemble JSON directly. I have no idea what library to use, but it seems like boost has support (https://www.boost.org/doc/libs/1_75_0/libs/json/doc/html/json/overview.html)

It's fine to not add the message for now, but please add a TODO to add the message and to use a JSON library.